File: UI/TemplateManagerWindow.py
import gi
import pathlib
gi.require_version("Gtk", "4.0")
from gi.repository import Gtk, Gdk, Gio, Pango
from UI.TemplateEditorWindow import TemplateEditorWindow
from UI.Icons import Icons
from Models.TemplateDataModel import TemplateDataModel

class TemplateManagerWindow(Gtk.ApplicationWindow):
    def __init__(self, parent_window, picker_mode=False, stream_type=None, on_select=None, **kwargs):
        super().__init__(**kwargs, title="Select Template" if picker_mode else "Manage Templates")
        self.app = Gtk.Application.get_default()
        self.picker_mode = picker_mode
        self.target_type = stream_type # e.g., 'video' or 'audio'
        self.on_select = on_select     # Callback for picker mode
        
        # Window Config
        self.set_default_size(750, 550)
        self.set_transient_for(parent_window)
        self.set_modal(True)

        # Header with Search
        hb = Gtk.HeaderBar()
        self.set_titlebar(hb)
        self.search_entry = Gtk.SearchEntry(placeholder_text="Filter templates...")
        self.search_entry.connect("search-changed", lambda e: self.populate_ui(e.get_text()))
        hb.set_title_widget(self.search_entry)

        # Main Layout
        main_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=8, margin_bottom=10, margin_end=10, margin_start=10, margin_top=10)
        self.set_child(main_box)

        # ListBox
        self.lst_templates = Gtk.ListBox()
        self.lst_templates.set_selection_mode(Gtk.SelectionMode.SINGLE)
        self.lst_templates.add_css_class("boxed-list")
        self.lst_templates.connect("selected-rows-changed", self.on_selection_changed)

        scroll = Gtk.ScrolledWindow(vexpand=True)
        scroll.set_child(self.lst_templates)
        main_box.append(scroll)

        # Toolbar / Action Box
        action_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6)
        main_box.append(action_box)

        self.btn_new = Gtk.Button(label="New Template", icon_name="list-add-symbolic")
        self.btn_new.set_tooltip_text("Create a brand new transcoding template")
        self.btn_new.connect("clicked", self.on_new_template)
        action_box.append(self.btn_new)

        self.btn_clone = Gtk.Button(label="Clone", icon_name="edit-copy-symbolic")
        self.btn_clone.set_tooltip_text("Create a copy of the selected template")
        self.btn_clone.set_sensitive(False)
        self.btn_clone.connect("clicked", self.on_clone_clicked)
        action_box.append(self.btn_clone)

        # Spacer to push Apply button to the right
        action_box.append(Gtk.Separator(orientation=Gtk.Orientation.HORIZONTAL, hexpand=True, opacity=0))

        if self.picker_mode:
            self.btn_apply = Gtk.Button(label="Apply Template")
            self.btn_apply.add_css_class("suggested-action")
            self.btn_apply.set_tooltip_text("Confirm selection and return to job setup")
            self.btn_apply.set_sensitive(False)
            self.btn_apply.connect("clicked", self.on_apply_clicked)
            action_box.append(self.btn_apply)

        self.populate_ui()

    def populate_ui(self, filter_text=""):
        while child := self.lst_templates.get_first_child():
            self.lst_templates.remove(child)
        
        # Logic: If target_type is set, filter by that type. Otherwise get all.
        if self.target_type:
            templates = TemplateDataModel.get_templates_by_type(self.app, self.target_type)
        else:
            templates = TemplateDataModel.get_all_templates(self.app)
        
        for t in templates:
            if not filter_text or filter_text.lower() in t["name"].lower():
                self.lst_templates.append(self._create_row(t))

    def _create_row(self, t):
        row_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6, margin_bottom=6, margin_end=6, margin_start=6, margin_top=6)
        
        # 1. Type Badge
        lbl_type = Gtk.Label(label=t["type"].upper())
        lbl_type.add_css_class("caption")
        lbl_type.set_width_chars(12)
        row_box.append(lbl_type)

        # 2. Icon
        icon_name = Icons.get_icon_for_type(t.get("type", "unknown").lower())
        row_box.append(Gtk.Image.new_from_icon_name(icon_name))

        # 3. Name (with Ellipsizing in the Middle)
        lbl_name = Gtk.Label(use_markup=True, xalign=0, hexpand=True)
        lbl_name.set_markup(f"<b>{t['name']}</b>")
        lbl_name.set_ellipsize(Pango.EllipsizeMode.MIDDLE) # Requirement met
        row_box.append(lbl_name)
        
        if t.get("readonly"):
            img_lock = Gtk.Image.new_from_icon_name("changes-prevent-symbolic")
            img_lock.set_tooltip_text("System Template")
            row_box.append(img_lock)

        # 4. Origin
        lbl_origin = Gtk.Label(label=t["origin"])
        lbl_origin.add_css_class("dim-label")
        row_box.append(lbl_origin)

        # 5. Buttons
        btn_edit = Gtk.Button(icon_name="document-edit-symbolic")
        btn_edit.connect("clicked", lambda *_: self.on_edit_clicked(t))
        row_box.append(btn_edit)

        # Metadata for row
        row = Gtk.ListBoxRow(child=row_box)
        row._data = t
        return row

    def on_selection_changed(self, _):
        row = self.lst_templates.get_selected_row()
        has_selection = row is not None
        self.btn_clone.set_sensitive(has_selection)
        if self.picker_mode:
            self.btn_apply.set_sensitive(has_selection)

    def on_new_template(self, _):
        new_data = TemplateDataModel.create_empty_template(self.target_type or "video")
        win = TemplateEditorWindow(self, template=new_data, on_save_callback=lambda _: self.populate_ui())
        win.present()

    def on_clone_clicked(self, _):
        row = self.lst_templates.get_selected_row()
        if not row: return
        
        win = TemplateEditorWindow(
            self, 
            template=row._data, 
            clone_mode=True, 
            on_save_callback=lambda _: self.populate_ui()
        )
        win.present()

    def on_edit_clicked(self, template):
        is_readonly = template.get("readonly", False)
        win = TemplateEditorWindow(
            self, 
            template=template, 
            clone_mode=is_readonly, 
            on_save_callback=lambda _: self.populate_ui(),
            locked_type=self.target_type
        )
        win.present()

    def on_apply_clicked(self, _):
        row = self.lst_templates.get_selected_row()
        if row and self.on_select:
            self.on_select(row._data['name'])
            self.destroy()

File: src/UI/TemplateManagerWindow.cs
using Gtk;
using FFGui.Models;
using FFGui.Core;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;
using Gio;
using File = System.IO.File;
using HarfBuzz;

namespace FFGui.UI;

public class TemplateManagerWindow : Window
{
    private class TemplateRow : ListBoxRow
    {
        public Template Template { get; init; } = null!;
    }

    private readonly FFGuiApp _app;
    private readonly Window _parent;
    private readonly bool _pickerMode;
    private readonly string _targetType;
    private readonly string _subType;
    private readonly Action<Template>? _onSelect;
    private readonly SizeGroup _sgType = new();
    private readonly SizeGroup _sgTarget = new();

    private ListBox _lstTemplates = null!;
    private SearchEntry _searchEntry = null!;
    private Button _btnApply = null!;
    private List<Template> _allTemplates = new();

    private readonly IDeserializer _yamlDeserializer;

    public TemplateManagerWindow(FFGuiApp app, Window parent) : this(app, parent, false, "", "", null) { }

    public TemplateManagerWindow(FFGuiApp app, Window parent, bool pickerMode, string targetType, string subType, Action<Template>? onSelect)
    {
        _app = app;
        _parent = parent;
        _pickerMode = pickerMode;
        _targetType = targetType;
        _subType = subType;
        _onSelect = onSelect;

        // Initialize YAML deserializer
        _yamlDeserializer = new DeserializerBuilder()
            .WithNamingConvention(PascalCaseNamingConvention.Instance)
            .IgnoreUnmatchedProperties()
            .Build();

        SetTitle(_pickerMode ? $"Select {_targetType} Template" : "Manage Templates");
        SetDefaultSize(800, 550);
        if (parent is not null) SetTransientFor(_parent);
        SetModal(true);

        BuildUI();
        LoadTemplates();
    }

    private void BuildUI()
    {
        var header = new HeaderBar();
        _searchEntry = new SearchEntry { PlaceholderText = "Filter templates..." };
        // Fix: Ensure we don't pass null to PopulateList
        _searchEntry.OnSearchChanged += (s, e) => PopulateList(_searchEntry.GetText() ?? "");
        header.SetTitleWidget(_searchEntry);
        SetTitlebar(header);

        var mainBox = new Box { Spacing = 8 };
        mainBox.SetOrientation(Orientation.Vertical);
        mainBox.SetMarginStart(12); mainBox.SetMarginEnd(12);
        mainBox.SetMarginTop(10); mainBox.SetMarginBottom(10);

        var listHeader = new Box { Spacing = 10, MarginStart = 8, MarginEnd = 12 };

        var lblType = new Label { Label_ = "<b>TYPE</b>", UseMarkup = true, Xalign = 0 };
        var lblTarget = new Label { Label_ = "<b>TARGET</b>", UseMarkup = true, Xalign = 0 };
        var lblName = new Label { Label_ = "<b>NAME</b>", UseMarkup = true, Xalign = 0, Hexpand = true };

        // Add to SizeGroups
        _sgType.Mode = SizeGroupMode.Horizontal;
        _sgType.AddWidget(lblType);
        _sgTarget.Mode = SizeGroupMode.Horizontal;
        _sgTarget.AddWidget(lblTarget);

        listHeader.Append(lblType);
        listHeader.Append(lblTarget);
        listHeader.Append(lblName);

        if (!_pickerMode)
        {
            // Align the Action header to the right
            listHeader.Append(new Label { Label_ = "<b>ACTIONS</b>", UseMarkup = true, Xalign = 1, WidthRequest = 120 });
        }
        mainBox.Append(listHeader);

        var scroll = new ScrolledWindow { Vexpand = true, HasFrame = true };
        _lstTemplates = new ListBox { SelectionMode = SelectionMode.Single };
        _lstTemplates.AddCssClass("boxed-list");
        _lstTemplates.OnSelectedRowsChanged += (s, e) => UpdateButtonSensitivity();
        _lstTemplates.OnRowActivated += (s, e) => { if (_pickerMode) _onApplyClicked(); };

        scroll.SetChild(_lstTemplates);
        mainBox.Append(scroll);

        var footer = new Box { Spacing = 6, Halign = Align.End };
        var btnCancel = new Button { Label = "Close" };
        btnCancel.OnClicked += (s, e) => Close();
        footer.Append(btnCancel);

        if (_pickerMode)
        {
            _btnApply = new Button { Label = "Apply Template" };
            _btnApply.AddCssClass("suggested-action");
            _btnApply.SetSensitive(false);
            btnCancel.Label = "Cancel";
            footer.Append(_btnApply);
            _btnApply.OnClicked += (s, e) => _onApplyClicked();
        }
        else
        {
            var btnNew = new MenuButton
            {
                IconName = "list-add-symbolic",
                TooltipText = "Create New Template"
            };

            // Load the menu from the string we just added to Menus.cs
            var builder = Builder.NewFromString(Menus.TemplateNewMenu, Menus.TemplateNewMenu.Length);
            
            var menuModel = builder.GetObject("template-new-menu") as MenuModel ?? throw new Exception("Menu model not found");
            btnNew.MenuModel = menuModel;

            var actionGroup = new SimpleActionGroup();
            // Fix: Follow JobRow's pattern for AddAction
            AddAction(actionGroup, "new_transcoding", (a, p) => _createNewTemplate(new TranscodingTemplate()));
            AddAction(actionGroup, "new_container", (a, p) => _createNewTemplate(new ContainerTemplate()));
            AddAction(actionGroup, "new_filter", (a, p) => _createNewTemplate(new FilterTemplate()));

            // Register the group under the prefix "tpl" (matching our XML)
            InsertActionGroup("tpl", actionGroup);

            footer.Append(btnNew);
        }

        mainBox.Append(footer);
        SetChild(mainBox);
    }

    private void LoadTemplates()
    {
        _allTemplates.Clear();

        foreach (var path in _app.TemplatePaths)
        {
            if (!Directory.Exists(path)) continue;

            // Adapt: Search for .yaml and .yml files
            string[] files = Directory.GetFiles(path, "*.y*ml");
            foreach (var file in files)
            {
                try
                {
                    string yaml = System.IO.File.ReadAllText(file);
                    Template? t = null;

                    // Polymorphic YAML detection based on content keys
                    if (yaml.Contains("EncoderSettings:"))
                        t = _yamlDeserializer.Deserialize<TranscodingTemplate>(yaml);
                    else if (yaml.Contains("Muxer:"))
                        t = _yamlDeserializer.Deserialize<ContainerTemplate>(yaml);
                    else if (yaml.Contains("Filters:"))
                        t = _yamlDeserializer.Deserialize<FilterTemplate>(yaml);

                    if (t != null)
                    {
                        // CAPTURE THE FILENAME HERE
                        t.Name = Path.GetFileNameWithoutExtension(file);
                        _allTemplates.Add(t);
                    }
                }
                catch { /* Log error or skip */ }
            }
        }
        PopulateList();
    }

    private void PopulateList(string filter = "")
    {
        while (_lstTemplates.GetFirstChild() is Widget child)
            _lstTemplates.Remove(child);

        var query = filter.ToLower();
        var filtered = _allTemplates.Where(t =>
            string.IsNullOrEmpty(query) ||
            t.Description.ToLower().Contains(query)
        );

        // Sort list
        filtered =
        [
            .. filtered
            .OrderBy(i => i switch
            {
                TranscodingTemplate _ => 0,
                FilterTemplate _      => 1,
                ContainerTemplate _   => 2,
                _ => 3
            })
            .ThenBy(i => i switch
            {
                TranscodingTemplate tt => tt.Type ?? "",
                ContainerTemplate ct   => ct.Muxer ?? "",
                FilterTemplate ft      => ft.Type ?? "",
                _ => ""
            }, StringComparer.OrdinalIgnoreCase)
            .ThenBy(i => i.Name ?? "", StringComparer.OrdinalIgnoreCase)
        ];


        foreach (var t in filtered)
        {
            if (_pickerMode)
            {
                if (t.GetType().ToString() != _targetType)
                    continue;

                if (t is TranscodingTemplate tt)
                    if (tt.Type != _subType)
                        continue;

                if (t is FilterTemplate ft)
                    if (ft.Type != _subType)
                        continue;
            }
            _lstTemplates.Append(CreateTemplateRow(t));
        }
    }

    private ListBoxRow CreateTemplateRow(Template t)
    {
        var row = new TemplateRow { Template = t }; // Tag the model here
        var box = new Box { Spacing = 10 };
        box.SetMarginStart(6); box.SetMarginEnd(6);
        box.SetMarginTop(4); box.SetMarginBottom(4);

        string typeStr = t switch
        {
            TranscodingTemplate => "TRANSCODING",
            ContainerTemplate => "CONTAINER",
            FilterTemplate => "FILTER",
            _ => "UNKNOWN"
        };

        string targetStr = t switch
        {
            TranscodingTemplate tt => tt.Type.ToUpper(),
            FilterTemplate ft => ft.Type.ToUpper(),
            ContainerTemplate ct => ct.Muxer.ToUpper(),
            _ => "N/A"
        };

        var lblType = new Label { Label_ = typeStr, Xalign = 0 };
        var lblTarget = new Label { Label_ = targetStr, Xalign = 0, Opacity = 0.7 };

        // Add to SizeGroups to sync width with Header
        _sgType.AddWidget(lblType);
        _sgTarget.AddWidget(lblTarget);

        box.Append(lblType);
        box.Append(lblTarget);

        box.Append(new Label { Label_ = t.Name, Xalign = 0, Hexpand = true, Ellipsize = Pango.EllipsizeMode.End });

        // --- Path Logic to detect Read-Only ---
        // In your app, index 0 is typically the system path, index 1 is user path
        bool isReadOnly = false;
        if (_app.TemplatePaths.Length > 1)
        {
            var systemPath = _app.TemplatePaths[0];
            // Check if the template file exists in the system directory
            var expectedSystemFile = Path.Combine(systemPath, $"{t.Name}.yaml");
            if (File.Exists(expectedSystemFile) && !_app.PortableMode) isReadOnly = true;
        }

        if (!_pickerMode)
        {
            var actionBox = new Box { Spacing = 4 };

            // 1. Move Clone Button into the row (Available for everyone)
            var btnCloneRow = Button.NewFromIconName("edit-copy-symbolic");
            btnCloneRow.TooltipText = "Clone Template";
            btnCloneRow.AddCssClass("flat");
            btnCloneRow.OnClicked += (s, e) => _onCloneClicked(t); // Modified to accept template
            actionBox.Append(btnCloneRow);

            // 2. Add Edit/Rename/Delete ONLY if not Read-Only
            if (!isReadOnly)
            {
                var btnEdit = Button.NewFromIconName("document-edit-symbolic");
                btnEdit.TooltipText = "Edit Template";
                btnEdit.AddCssClass("flat");
                btnEdit.OnClicked += (s, e) => _onEditClicked(t);
                actionBox.Append(btnEdit);

                var btnRename = Button.NewFromIconName("insert-text-symbolic");
                btnRename.TooltipText = "Rename Template";
                btnRename.AddCssClass("flat");
                btnRename.OnClicked += (s, e) => _onRenameClicked(t);
                actionBox.Append(btnRename);

                var btnDelete = Button.NewFromIconName("user-trash-symbolic");
                btnDelete.TooltipText = "Delete Template";
                btnDelete.AddCssClass("flat");
                btnDelete.AddCssClass("destructive-action");
                btnDelete.OnClicked += (s, e) => _onDeleteClicked(t);
                actionBox.Append(btnDelete);
            }
            else
            {
                // Optional: Add a lock icon for visual feedback
                var lblLocked = Image.NewFromIconName("changes-prevent-symbolic");
                lblLocked.TooltipText = "System Template (Read-Only)";
                lblLocked.Opacity = 0.5;
                actionBox.Append(lblLocked);
            }

            box.Append(actionBox);
        }

        row.SetChild(box);
        return row;
    }

    private void UpdateButtonSensitivity()
    {
        bool hasSelection = _lstTemplates.GetSelectedRow() != null;
        if (_pickerMode) _btnApply.SetSensitive(hasSelection);
    }

    private void AddAction(SimpleActionGroup group, string name, Action<SimpleAction, GLib.Variant?> callback)
    {
        var action = SimpleAction.New(name, null);
        action.OnActivate += (s, e) => callback(s, e.Parameter);
        group.AddAction(action);
    }

    private void _createNewTemplate(Template tpl)
    {
        // Launch editor in 'New' mode
        var editor = new TemplateEditorWindow(_app, tpl, EditorMode.New, (savedTpl, name) =>
        {
            _saveTemplateToDisk(savedTpl, name);
            LoadTemplates(); // Refresh the list
        });
        editor.SetTransientFor(this);
        editor.Present();
    }

    private void _saveTemplateToDisk(Template tpl, string name)
    {

        if (_app.TemplatePaths == null)
        {
            Console.WriteLine("Error: User template path (index 1) is not configured.");
            return;
        }
        var userPath = "";
        if (_app.TemplatePaths.Length == 1)
        {
            // Portable mode
            userPath = _app.TemplatePaths[0];
        }
        else
        {
            // Installed mode
            userPath = _app.TemplatePaths[1];
        }

        if (!Directory.Exists(userPath)) Directory.CreateDirectory(userPath);

        // Use the TemplateName set by the editor
        var filePath = Path.Combine(userPath, $"{name}.yaml");

        var serializer = new SerializerBuilder()
            .WithNamingConvention(PascalCaseNamingConvention.Instance)
            .Build();

        var yaml = serializer.Serialize(tpl);
        System.IO.File.WriteAllText(filePath, yaml);
    }

    private void _onDeleteClicked(Template? t = null)
    {
        var selectedTemplate = t ?? _getSelectedTemplate();
        if (selectedTemplate == null) return;

        // FIX: Handle both Installed (Length=2) and Portable (Length=1) modes
        string userPath;
        if (_app.TemplatePaths.Length > 1)
        {
            // Installed Mode: User templates are at index 1
            userPath = _app.TemplatePaths[1];
        }
        else if (_app.TemplatePaths.Length == 1)
        {
            // Portable Mode: Templates are at index 0
            userPath = _app.TemplatePaths[0];
        }
        else
        {
            return; // Should not happen
        }

        // Use the Name property we added to the model
        var filePath = Path.Combine(userPath, $"{selectedTemplate.Name}.yaml");

        if (System.IO.File.Exists(filePath))
        {
            System.IO.File.Delete(filePath);
            LoadTemplates();
        }
        else
        {
            Console.WriteLine($"Cannot delete '{selectedTemplate.Name}'. It might be a system template or read-only.");
        }
    }

    private void _onRenameClicked(Template t)
    {
        // Simple input dialog using a Window
        var dialog = new Window { Title = "Rename Template", Modal = true, TransientFor = this };
        dialog.SetDefaultSize(300, 100);

        var box = new Box { Spacing = 10, MarginTop = 12, MarginBottom = 12, MarginStart = 12, MarginEnd = 12 };
        box.SetOrientation(Orientation.Vertical);

        var entry = new Entry { Text_ = t.Name };
        Entry.ActivateSignal.Connect(entry, (sender, e) =>
        {
            _performRename(t, entry.GetText());
            dialog.Close();
        });

        var btnBox = new Box { Spacing = 6, Halign = Align.End };
        var btnCancel = new Button { Label = "Cancel" };
        btnCancel.OnClicked += (s, e) => dialog.Close();

        var btnSave = new Button { Label = "Rename" };
        btnSave.AddCssClass("suggested-action");
        btnSave.OnClicked += (s, e) =>
        {
            _performRename(t, entry.GetText());
            dialog.Close();
        };

        btnBox.Append(btnCancel);
        btnBox.Append(btnSave);

        box.Append(new Label { Label_ = "Enter new name:", Xalign = 0 });
        box.Append(entry);
        box.Append(btnBox);

        dialog.SetChild(box);
        dialog.Present();
    }

    private void _onEditClicked(Template? t = null)
    {
        // If called from the ListBox activation (null), get selected
        var selectedTemplate = t ?? _getSelectedTemplate();
        if (selectedTemplate == null) return;

        var editor = new TemplateEditorWindow(_app, selectedTemplate, EditorMode.Edit, (savedTpl, name) =>
        {
            _saveTemplateToDisk(savedTpl, name);
            LoadTemplates();
        });
        editor.SetTransientFor(this);
        editor.Present();
    }

    private void _onCloneClicked(Template? t = null)
    {
        var selectedTemplate = t ?? _getSelectedTemplate();
        if (selectedTemplate == null) return;

        var serializer = new SerializerBuilder().WithNamingConvention(PascalCaseNamingConvention.Instance).Build();
        var yaml = serializer.Serialize(selectedTemplate);

        Template? clonedTemplate = null;
        if (selectedTemplate is TranscodingTemplate) clonedTemplate = _yamlDeserializer.Deserialize<TranscodingTemplate>(yaml);
        else if (selectedTemplate is ContainerTemplate) clonedTemplate = _yamlDeserializer.Deserialize<ContainerTemplate>(yaml);
        else if (selectedTemplate is FilterTemplate) clonedTemplate = _yamlDeserializer.Deserialize<FilterTemplate>(yaml);

        if (clonedTemplate != null)
        {
            clonedTemplate.Description += " (Clone)";
            _createNewTemplate(clonedTemplate);
        }
    }

    private void _onDeleteClicked()
    {
        var selectedTemplate = _getSelectedTemplate();
        if (selectedTemplate == null) return;

        // Warning: System templates (Path index 0) should probably be read-only
        // We only delete from the User path (Path index 1)
        string safeName = string.Concat(selectedTemplate.Description.Split(Path.GetInvalidFileNameChars())).Replace(" ", "_");
        var userFilePath = Path.Combine(_app.TemplatePaths[1], $"{safeName}.yaml");

        if (System.IO.File.Exists(userFilePath))
        {
            System.IO.File.Delete(userFilePath);
            LoadTemplates();
        }
        else
        {
            // Optional: show a message that system templates cannot be deleted
            Console.WriteLine("Cannot delete system template.");
        }
    }

    private void _onApplyClicked()
    {
        var selectedTemplate = _getSelectedTemplate();
        if (selectedTemplate != null && _onSelect != null)
        {
            _onSelect(selectedTemplate);
            Close();
        }
    }

    private void _performRename(Template t, string newName)
    {
        newName = newName.Trim();
        if (string.IsNullOrEmpty(newName) || newName == t.Name) return;

        // Determine user path safely
        var userPath = _app.TemplatePaths.Length > 1 ? _app.TemplatePaths[1] : _app.TemplatePaths[0];

        var oldPath = Path.Combine(userPath, $"{t.Name}.yaml");
        var newPath = Path.Combine(userPath, $"{newName}.yaml");

        if (File.Exists(newPath))
        {
            Console.WriteLine("Error: A template with that name already exists.");
            return;
        }

        if (File.Exists(oldPath))
        {
            try
            {
                File.Move(oldPath, newPath);
                t.Name = newName; // Update model
                LoadTemplates(); // Refresh UI
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Rename failed: {ex.Message}");
            }
        }
        else
        {
            Console.WriteLine("Error: Cannot find original template file. It might be a system template.");
        }
    }

    // Helper to get the template model from the selected ListBoxRow
    private Template? _getSelectedTemplate()
    {
        if (_lstTemplates.GetSelectedRow() is TemplateRow tRow)
        {
            return tRow.Template;
        }
        return null;
    }
}